Python 3.7 test failures

Running tests from master under Python 3.7.1 (which isn't tested on Travis) I see some failures. The fastest ones to see are from `test_implicit_multiplication_application.py` but there is another in `test_sympify.py`. I'm not sure what has changed in 3.7 to cause this.

```console
$ ./bin/test sympy/parsing/tests/test_implicit_multiplication_application.py 
==================================================================== test process starts =====================================================================
executable:         /Users/enojb/current/sympy/upstream.sympy.git/venv371/bin/python  (3.7.1-final-0) [CPython]
architecture:       64-bit
cache:              yes
ground types:       python 
numpy:              None
random seed:        88174423
hash randomization: on (PYTHONHASHSEED=224844549)

sympy/parsing/tests/test_implicit_multiplication_application.py[5] .E..E                                                                                [FAIL]

______________________________________________________________________________________________________________________________________________________________
_________________________________ sympy/parsing/tests/test_implicit_multiplication_application.py:test_implicit_application __________________________________
Traceback (most recent call last):
  File "/Users/enojb/current/sympy/upstream.sympy.git/sympy/core/cache.py", line 94, in wrapper
    retval = cfunc(*args, **kwargs)
  File "/Users/enojb/current/sympy/upstream.sympy.git/sympy/core/function.py", line 446, in __new__
    'given': n})
TypeError: factorial takes exactly 1 argument (0 given)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/enojb/current/sympy/upstream.sympy.git/sympy/parsing/tests/test_implicit_multiplication_application.py", line 62, in test_implicit_application
    implicit = parse_expr(case, transformations=transformations2)
  File "/Users/enojb/current/sympy/upstream.sympy.git/sympy/parsing/sympy_parser.py", line 965, in parse_expr
    return eval_expr(code, local_dict, global_dict)
  File "/Users/enojb/current/sympy/upstream.sympy.git/sympy/parsing/sympy_parser.py", line 878, in eval_expr
    code, global_dict, local_dict)  # take local objects in preference
  File "<string>", line 1, in <module>
  File "/Users/enojb/current/sympy/upstream.sympy.git/sympy/core/cache.py", line 96, in wrapper
    retval = func(*args, **kwargs)
  File "/Users/enojb/current/sympy/upstream.sympy.git/sympy/core/function.py", line 446, in __new__
    'given': n})
TypeError: factorial takes exactly 1 argument (0 given)
______________________________________________________________________________________________________________________________________________________________
__________________________________ sympy/parsing/tests/test_implicit_multiplication_application.py:test_all_implicit_steps ___________________________________
Traceback (most recent call last):
  File "/Users/enojb/current/sympy/upstream.sympy.git/sympy/core/cache.py", line 94, in wrapper
    retval = cfunc(*args, **kwargs)
  File "/Users/enojb/current/sympy/upstream.sympy.git/sympy/core/function.py", line 446, in __new__
    'given': n})
TypeError: factorial takes exactly 1 argument (0 given)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/enojb/current/sympy/upstream.sympy.git/sympy/parsing/tests/test_implicit_multiplication_application.py", line 183, in test_all_implicit_steps
    implicit = parse_expr(case, transformations=transformations2)
  File "/Users/enojb/current/sympy/upstream.sympy.git/sympy/parsing/sympy_parser.py", line 965, in parse_expr
    return eval_expr(code, local_dict, global_dict)
  File "/Users/enojb/current/sympy/upstream.sympy.git/sympy/parsing/sympy_parser.py", line 878, in eval_expr
    code, global_dict, local_dict)  # take local objects in preference
  File "<string>", line 1, in <module>
  File "/Users/enojb/current/sympy/upstream.sympy.git/sympy/core/cache.py", line 96, in wrapper
    retval = func(*args, **kwargs)
  File "/Users/enojb/current/sympy/upstream.sympy.git/sympy/core/function.py", line 446, in __new__
    'given': n})
TypeError: factorial takes exactly 1 argument (0 given)

================================================== tests finished: 3 passed, 2 exceptions, in 0.26 seconds ===================================================
DO *NOT* COMMIT!
```

The error from `test_sympify.py` is
```console
========================================================================== FAILURES ==========================================================================
___________________________________________________________________ test_sympify_keywords ____________________________________________________________________

    def test_sympify_keywords():
        raises(SympifyError, lambda: sympify('if'))
        raises(SympifyError, lambda: sympify('for'))
        raises(SympifyError, lambda: sympify('while'))
>       raises(SympifyError, lambda: sympify('lambda'))

sympy/core/tests/test_sympify.py:148: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
sympy/core/tests/test_sympify.py:148: in <lambda>
    raises(SympifyError, lambda: sympify('lambda'))
sympy/core/sympify.py:372: in sympify
    expr = parse_expr(a, local_dict=locals, transformations=transformations, evaluate=evaluate)
sympy/parsing/sympy_parser.py:965: in parse_expr
    return eval_expr(code, local_dict, global_dict)
sympy/parsing/sympy_parser.py:878: in eval_expr
    code, global_dict, local_dict)  # take local objects in preference
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

>   ???
E   TypeError: __new__() missing 1 required positional argument: 'expr'

<string>:1: TypeError
```

--- a/sympy/parsing/sympy_parser.py
+++ b/sympy/parsing/sympy_parser.py
@@ -3,7 +3,7 @@
 from __future__ import print_function, division
 
 from tokenize import (generate_tokens, untokenize, TokenError,
-    NUMBER, STRING, NAME, OP, ENDMARKER, ERRORTOKEN)
+    NUMBER, STRING, NAME, OP, ENDMARKER, ERRORTOKEN, NEWLINE)
 
 from keyword import iskeyword
 
@@ -262,9 +262,7 @@ def _implicit_application(tokens, local_dict, global_dict):
                           # work with function exponentiation
     for tok, nextTok in zip(tokens, tokens[1:]):
         result.append(tok)
-        if (tok[0] == NAME and
-              nextTok[0] != OP and
-              nextTok[0] != ENDMARKER):
+        if (tok[0] == NAME and nextTok[0] not in [OP, ENDMARKER, NEWLINE]):
             if _token_callable(tok, local_dict, global_dict, nextTok):
                 result.append((OP, '('))
                 appendParen += 1
@@ -570,8 +568,11 @@ def lambda_notation(tokens, local_dict, global_dict):
     flag = False
     toknum, tokval = tokens[0]
     tokLen = len(tokens)
+
     if toknum == NAME and tokval == 'lambda':
-        if tokLen == 2:
+        if tokLen == 2 or tokLen == 3 and tokens[1][0] == NEWLINE:
+            # In Python 3.6.7+, inputs without a newline get NEWLINE added to
+            # the tokens
             result.extend(tokens)
         elif tokLen > 2:
             result.extend([

--- a/sympy/parsing/tests/test_implicit_multiplication_application.py
+++ b/sympy/parsing/tests/test_implicit_multiplication_application.py
@@ -61,7 +61,7 @@ def test_implicit_application():
     for case in cases:
         implicit = parse_expr(case, transformations=transformations2)
         normal = parse_expr(cases[case], transformations=transformations)
-        assert(implicit == normal)
+        assert(implicit == normal), (implicit, normal)
 
     multiplication = ['x y', 'x sin x', '2x']
     for case in multiplication:

